src/wayland/dmabuf/mod.rs

```diff
@@ -112,18 +119,311 @@ use crate::{
+        };
+
+        Self {
+            main_device,
+            formats: feedback_formats,
+            main_tranche,
+            preferred_tranches: Vec::new(),
+        }
+    }
+
+    /// Adds a preferred tranche to the builder
+    ///
+    /// The tranches will be reported in the order they have been added with
+    /// this function.
+    ///
+    /// Note: Formats already present in a previously added preferred tranche with the
+    /// same target device and flags will be skipped.
+    /// If all formats are already included in a previously added tranche
+    /// with the same target device and flags this tranche will be skipped.
+    pub fn add_preferred_tranche(
+        mut self,
+        target_device: libc::dev_t,
+        flags: Option<zwp_linux_dmabuf_feedback_v1::TrancheFlags>,
+        formats: impl IntoIterator<Item = Format>,
+    ) -> Self {
+        let flags = flags.unwrap_or(zwp_linux_dmabuf_feedback_v1::TrancheFlags::empty());
+
+        let mut tranche = DmabufFeedbackTranche {
+            target_device,
+            flags,
+            indices: Default::default(),
+        };
+
+        for format in formats {
+            let (format_index, added) = self.formats.insert_full(format);
+
+            // Compositors must not send duplicate format + modifier pairs within
+            // the same tranche or across two different tranches with the same target device and flags.
+            //
+            // if the format has just been added there is no need to test if a previous tranche
+            // with the same target device and flags already contains the format
+            let duplicate_format = !added
+                && self.preferred_tranches.iter().any(|tranche| {
+                    tranche.target_device == target_device
+                        && tranche.flags == flags
+                        && tranche.indices.contains(&format_index)
+                });
+
+            if duplicate_format {
+                continue;
+            }
+
+            // ...Compositors must not send duplicate format + modifier pairs within the same tranche...
+            // this is handled by using a IndexSet which won't hold duplicates
+            tranche.indices.insert(format_index);
+        }
+
+        if !tranche.indices.is_empty() {
+            self.preferred_tranches.push(tranche);
+        }
+
+        self
+    }
+
+    /// Build the [`DmabufFeedback`]
+    ///
+    /// Returns an error if the format table shared memory file could
+    /// not be created.
+    pub fn build(mut self) -> Result<DmabufFeedback, std::io::Error> {
+        let formats = self
+            .formats
+            .iter()
+            .copied()
+            .map(DmabufFeedbackFormat::from)
+            .collect::<Vec<_>>();
+
+        let name = CString::new("smithay-dmabuffeedback-format-table").unwrap();
+        let format_table_file = SealedFile::with_data(name, unsafe {
+            std::slice::from_raw_parts(
```

Comment:
I think we could do this without unsafe. We can use `u32::into_bytes_ne`, and then collect all the u8 bytes.

Reply:
added `DmabufFeedbackFormat::to_ne_bytes` and replaced the unsafe part with `.flat_map(DmabufFeedbackFormat::to_ne_bytes).collect()`